### ISSUE TITLE
Testing left to do.

### ISSUE THREAD

**Issue by rgleason** [jongough/testplugin_pi#10]
1. Get the ocpn-plugin.xml built and working.
2. Test the operation of ocpn-plugin.xml for each OS.
3. Possibly fix some small items in testplugin_pi  #9

**Comment by jongough** [comment]
1. Is this an issue for testplugin_pi or is it an issue for the new installation process? The xml that is generated by this plugin does work if added to the ocpn-plugin.xml (I currently do it by hand as it is quicker and easier than the current automated process).

2. I think the testing should be to see if the testplugin_pi works for each OS rather than the xml file. I currently have a few issues with the 'working' bit.

3.  I am trying to find what the items that are not working are. 
   a. I know that the icon delivery for the plugin for windows is not working.
   b. I know that there is an issue under ubuntu 16.04 where the installation process works but OCPN crashes as soon as the plugin is enabled (this happens with squiddio as well). It appears that the process may not play well if a previous version of OCPN was installed.
   c. I know that there are issues, at times, with the build process in circleci failing, but this is out of my control.

Perhaps there should be a forum stream running for this?

**Comment by rgleason** [comment]
I find the ocpn-plugins batch files do not work and was starting to work it out, but Leamas has made a new PR now. https://github.com/OpenCPN/plugins/pull/9   so I will try that.

**Comment by rgleason** [comment]
Using the PI Installer  > Advanced > URL 
![Custom-Catalog-URL](https://user-images.githubusercontent.com/2363959/70005965-6b8ddd80-1539-11ea-87c1-c13d997cf628.jpg)

But the path is not quite right so it fails..
![Failure-wrong-path](https://user-images.githubusercontent.com/2363959/70005818-dee31f80-1538-11ea-9d06-7195954da224.jpg)

Looks like  ......"//dl.cloudsmith.io/public/rick-gleason/testplugin/raw/names/testplugin-msvc-10.0.14393-tarball/...

When I look at the ocpn-plugins.xml I find this path.
```
<tarball-url>
https://dl.cloudsmith.io/public/rick-gleason/testplugin/raw/names/testplugin-msvc-10.0.14393-tarball/versions/v1.0.41.5/testplugin_pi-1.0.41.5-ov50-1.16_msvc-10.0.14393.tar.gz
</tarball-url>
```
Failed
![Failure](https://user-images.githubusercontent.com/2363959/70006051-c6bfd000-1539-11ea-8123-9bd95c7eb5f9.jpg)

Also notice that the long description for testplugin screws up the menu. The description should be shortened IMHO.

**Comment by rgleason** [comment]
When I try this path used in the ocpn-plugins.xml it fails.
https://dl.cloudsmith.io/public/rick-gleason/testplugin/raw/names/testplugin-msvc-10.0.14393-tarball/versions/v1.0.41.5/testplugin_pi-1.0.41.5-ov50-1.16_msvc-10.0.14393.tar.gz

The path should be 
https://dl.cloudsmith.io/public/rick-gleason/opencpn-plugins-stable/raw/names/testplugin-msvc-10.0.14393-tarball/versions/v1.0.41.6/testplugin_pi-1.0.41.6-ov50-1.16_msvc-10.0.14393.tar.gz
Which works.

**Comment by rgleason** [comment]
Jon,
Creation of the final pathname for the xml is not taking into account where my Cloudsmith files are actually located.

Please see how I have had to change the upload files.

https://github.com/rgleason/testplugin_pi/blob/cmake_flatpak_test/cmake/in-files/appveyor-upload.sh.in

https://github.com/rgleason/testplugin_pi/blob/cmake_flatpak_test/cmake/in-files/circleci-upload.sh.in

Note these changes:

```
#STABLE_REPO='@CLOUDSMITH_USER@/@GIT_REPOSITORY_NAME@-stable'
#UNSTABLE_REPO='@CLOUDSMITH_USER@/@GIT_REPOSITORY_NAME@'
#STABLE_PKG_REPO='@CLOUDSMITH_USER@/@GIT_REPOSITORY_NAME@-pkg'

STABLE_REPO='@CLOUDSMITH_USER@/opencpn-plugins-stable'
UNSTABLE_REPO='@CLOUDSMITH_USER@/opencpn-plugins-unstable'
STABLE_PKG_REPO='@CLOUDSMITH_USER@/opencpn-plugins-pkg'

```

**Comment by rgleason** [comment]
https://dl.cloudsmith.io/public/rick-gleason/testplugin/raw/**_names_**/...
VS
https://dl.cloudsmith.io/public/rick-gleason/**_opencpn-plugins-stable_**/raw/names/

**Comment by jongough** [comment]
The issue with this is that each person that 'builds' the plugin will have to modify the CMakeLists.txt file to reflect the cloudsmith repository which will mean having multiple versions of the file around that should not be submitted as a pull request.

This comes down to naming and what/where these objects are going to be stored. If it is the developer cloudsmiths then you will not have your version. If it is a global cloudsmith then the file can be changed to reflect this and each change will be reflected there.

Currently the name of the cloudsmith repository is built from the GIT_REPOSITORY_NAME, so your cloudsmith breaks this. I have just pushed a change which allows you to explicitly set the cloudsmith base respository name (i.e. without -stable or -manual, suffices).

**Comment by rgleason** [comment]
> The issue with this is that each person that 'builds' the plugin will have to modify the CMakeLists.txt file to reflect the cloudsmith repository which will mean having multiple versions of the file around that should not be submitted as a pull request.

Jon, I had not thought about this point.  For the most part, you've consolidated the specific changes to the top of CMakeLists.txt which is great.  Won't that person that wants to build the plugin need to modify the top of CMakeLists.txt anyway? 

And if they wish to deploy to cloudsmith some special version of the plugin, they will then have to set the cloudsmith directory.  For example, Petri's joining of Dashboard and Tactics which he now calls Dashboard-Tactics_pi, I guess he would just have to create & enable the correct Cloudsmith repositories ( dashboard-tactics, dashboard-tactics-stable, dashboard-tactics-manual ) and it would all work very nicely.  I suppose that could be considered to be the standard distributed approach.

Have we decided that each and every plugin with have its own self named repository?  Alec has advised that Cloudsmith has good search features and I have found the [Package-Groups"](https://cloudsmith.io/~rick-gleason/repos/opencpn-plugins-pkg/groups/) view to be very very helpful, particularly for manual download from  the "manual" repository. 

Also Alec has indicated that all plugins could be stored in a less distributed way, perhaps centrally, using repositories named  (opencpn-plugins, opencpn-plugins-stable, opencpn-plugins-manual)  OR even (plugins, plugins-stable, plugins-manual). <---What do you think about the shorter version?

What happens if we provide a central Cloudsmith "Organization" named "OpenCPN" with the following repositories (plugins, plugins-stable, plugins-manual)?  Not completely sure how Organizations work, but I sent Lee's write up earlier.  I am    

Should I try setting up the "OpenCPN" Organization and inviting you? Should I create these repositories (plugins, plugins-stable, plugins-manual)?  OR would it be better to have 3 x # of plugins repositories that we as administrators, have to set up each time?  IE: testplugin, testplugin-stable, testplugin-manual.   Then we will need to give other plugin Dev's the Cloudsmith_API_Key to put into their Circleci and Appveyor.

**Comment by rgleason** [comment]
PS  Jon, I am going to remove some of my repositories and recreate & enable them to conform to the way you have designated the suffix, but still using opencpn-plugins:

- opencpn-plugins              (instead of using unstable)
- opencpn-plugins-stable
- opencpn-plugins-manual   (instead of using pkg)

 when not using the more centralized Organization repositories, I think the PI Installer Wiki should require PI Devs to make their repositories using [plugin_name]-[suffix as shown above].  Do you agree?

**Comment by rgleason** [comment]
Now I see your using "pkg" instead of "manual"

**Comment by jongough** [comment]
I am just following not leading on this. I am trying to make my packages match what I am seeing with Alec. It is very easy to change the suffix, I just need to know what to change it to and then have that stabilised. Now that my mods are working, i.e. deploying and findable, making name changes is not a problem. Again, I just need to know the stable naming that is going to be used. The reason I have all of them in testplugin is that I was testing to make sure the build and install process worked. I will delete the ones that should not be used.

As for names I would have through the suffices would have been:
dev or beta - is the current unstable
prod or released - is the current stable
pkg - is the current manual

I think pkg is better than manual as in the released, old stable, repository some of the xml refers to this repository for the actual installable objects. For naming I think we need to make it easy for users to understand what is in each repository, so naming could be important.

**Comment by rgleason** [comment]
I agree.  Ok, lets try to decide and then sell it to Alec.

Now                Use
unstable ---> beta
stable     ---> prod
manual ----> pkg

I think your suggestions are good. They're short and understandable.

**Comment by rgleason** [comment]
Alec can we get your agreement on these changes to the repository endings?
How would it affect the other parts of PI Installer?

- Now   ---->   Use
- unstable ---> beta
- stable ---> prod
- manual ----> pkg

Also these can be used as follows:

-     [pluginname]-beta
-     [pluginname]-prod
-     [pluginname]-pkg
-     Perhaps shorter and better if we did not require the "_pi" in the repository name. Do you agree?

For repositories that host more than one plugin.

-     opencpn-plugins-beta
-     opencpn-plugins-prod
-     opencpn-plugins-pkg

OR

-     plugins-beta
-     plugins-prod
-     plugins-pkg
-     If used with an organization "OpenCPN"

It will mean we will need to change our Cloudsmith repositories.

**Comment by leamas** [comment]
I'm agnostic about this, use whatever you think is best. IMHO different plugin projects will probably end up in different solutions on this.

**Comment by rgleason** [comment]
Thanks  Alec.  Jon, how about changing the suffixes to 

[pluginname_pi]-beta
[pluginname_pi]-prod
[pluginname_pi]-pkg

In thinking about the "_pi"  further, I think that need for the git the variable
set(GIT_REPOSITORY_NAME "testplugin_pi")
which is used elsewhere, makes it easiest to keep the "_pi" in the CS repositories.

For repositories that I am going to use for multiple PI
    opencpn-plugins-beta
    opencpn-plugins-prod
    opencpn-plugins-pkg

**Comment by rgleason** [comment]
1. Get the ocpn-plugin.xml built and working.  - It works now.
2. Test the operation of ocpn-plugin.xml for each OS.   
           (not done completely yet, will need beta testers.)
3. Possibly fix some small items in testplugin_pi #9  - Jon has fixed it.

Note, I'll change my repositories and the upload.sh.in  to go to
- opencpn-plugins-pkg
- opencpn-plugins-beta
- opencpn-plugins-prod

Also I'll going to review and rewrite the PI Installer wiki accordingly.
Also I will add and rewrite  "Testing"

Beta testers will need to load and test with their OS version of Opencpn with PI Installer to my https://github.com/rgleason/plugins  using the beta-ocpn-plugins.xml
URL: https://raw.githubusercontent.com/rgleason/plugins/master/beta-ocpn-plugins.xml

I do not intend to be the one who tests all the PI OS versions, instead using others to do that, and the individual plugin OS's  metadata should not be added to ocpn-plugins.xml until tested!

**Comment by rgleason** [comment]
I've reviewed and modified these pages for documentation:

[PI Installer Dev Setup](https://opencpn.org/wiki/dokuwiki/doku.php?id=opencpn:developer_manual:pi_installler_build_deploy)

[PI Installer Dev Procedure](https://opencpn.org/wiki/dokuwiki/doku.php?id=opencpn:developer_manual:pi_installer_procedure)

**Comment by jongough** [comment]
I have now managed to start the testing of a non 'core' plugin, i.e. testplugn_pi, as it now deploys to cloudsmith and for debian at least will generate a *.tar.gz installable that contains debugging information (I think). 

Having got this going I have raised a few new issues on Alec's pull request that need fixing in his code.

I will have a look at your doco soon, I hope.

**Comment by leamas** [comment]
This is great work. In particular, the overall description is spot on and very good.

Some small nags:

  - You need to emphasize the status of this thing: it's just a PR. Nobody can say for sure if/when this it is gonna be merged.

  - The CI parts, while useful,  are not necessary to use the installer. As long as there is a working file with metadata and a downloadable tarball everything is fine, at least for a starter. Important since the CI parts definitely  is an uphill battle.
  - You could need some references, including the wiki at  https://github.com/leamas/OpenCPN/wiki/

**Comment by rgleason** [comment]
I am hoping it will be merged! but will add all your points and the link, which looks very good. Thank you.

I am hoping to use the Cmake and CI files with weather_routing_pi first, once  Jon thinks it is ready. The rest of Sean's plugins will follow fairly soon once that is working.  I am going to have to find an army of OS Beta Testers to help.

**Comment by jongough** [comment]
I am not sure what has happened but website appears to be done. I did get a quick look at it yesterday and it seemed fine. I was going to have a closer look today, but all I get for any manual at the moment is the following:
![image](https://user-images.githubusercontent.com/2803240/70385047-b8603100-19dd-11ea-9771-989d31c26399.png)

**Comment by rgleason** [comment]
Website is fixed now. Had to revert to older bootstrap3 template.

**Comment by jongough** [comment]
Looks good. I did make a couple of small changes for clarity.

**Comment by rgleason** [comment]
I think I can close this now.
